### PR TITLE
[ONNX] propagate model directory path to Graph, Attribute and Tensor …

### PR DESCRIPTION
**Pull request by mateusztabaka**
…class

Required by external data mechanism in ONNX standard, where Tensor object has to be able to find external data file based on the location field and model directory path. Previously, it was done by a transformation, but it handled initializers only, but in order to handle Constant nodes we need one more loop, but over all of the model's nodes. Propagating model directory path to Tensor allows us to reduce that overhead.

Ticket: 91271

### CODE REVIEW

File: src/frontends/onnx/frontend/src/core/attribute.hpp [hunk 1]
```diff
@@ -299,11 +261,17 @@ class Attribute {
     Subgraph get_subgraph(const Graph* parent_graph) const;
 
     std::vector<Tensor> get_tensor_array() const {
-        return {std::begin(m_attribute_proto->tensors()), std::end(m_attribute_proto->tensors())};
+        std::vector<Tensor> ret;
```
Comment (t-jankowski) [inline src/frontends/onnx/frontend/src/core/attribute.hpp:264]: Vector size is known up front - consider `reserve()`.

File: src/frontends/onnx/frontend/src/utils/tensor_external_data.cpp [hunk 2]
```diff
@@ -17,26 +17,33 @@ namespace onnx_import {
...
+            NGRAPH_SUPPRESS_DEPRECATED_END
+        } else if (entry.key() == "offset") {
             m_offset = std::stoi(entry.value());
-        if (entry.key() == "length")
+        } else if (entry.key() == "length") {
             m_data_length = std::stoi(entry.value());
-        if (entry.key() == "checksum")
+        } else if (entry.key() == "checksum") {
             m_sha1_digest = std::stoi(entry.value());
+        }
```
Comment (t-jankowski) [inline src/frontends/onnx/frontend/src/utils/tensor_external_data.cpp:30]: Consider `switch` statement.

**Review by t-jankowski — CHANGES_REQUESTED** [review]

File: src/frontends/onnx/frontend/src/utils/tensor_external_data.cpp (hunk 2 above)
Comment (mateusztabaka) [reply src/frontends/onnx/frontend/src/utils/tensor_external_data.cpp:30]: you can't use switch with string in c++

Comment (t-jankowski) [reply src/frontends/onnx/frontend/src/utils/tensor_external_data.cpp:30]: Right .. may I delete my comment? ;)

Comment (mateusztabaka) [reply src/frontends/onnx/frontend/src/utils/tensor_external_data.cpp:30]: sure

File: src/frontends/onnx/frontend/src/core/tensor.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+template <>
+std::vector<int16_t> Tensor::get_data() const {
+    if (has_external_data()) {
+        return get_external_data<int16_t>();
+    }
+    if (m_tensor_proto->has_raw_data()) {
+        return detail::__get_raw_data<int16_t>(m_tensor_proto->raw_data(), m_tensor_proto->data_type());
+    }
+    if (m_tensor_proto->data_type() == ONNX_NAMESPACE::TensorProto_DataType_INT16) {
+        return detail::__get_data<int16_t>(m_tensor_proto->int32_data());
```
Comment (mbencer) [inline src/frontends/onnx/frontend/src/core/tensor.cpp:99]: shouldn't be `m_tensor_proto->int16_data`?

**Review by mbencer — APPROVED** [review]
LGTM

File: src/frontends/onnx/tests/onnx_import_external_data.in.cpp [hunk 4]
```diff
@@ -138,3 +138,16 @@ NGRAPH_TEST(${BACKEND_NAME}, onnx_external_data_sanitize_path) {
 
     test_case.run();
 }
+
+NGRAPH_TEST(${BACKEND_NAME}, onnx_external_data_in_constant_node) {
```
[the remark this answers is not in the document]
Comment (mateusztabaka) [reply src/frontends/onnx/tests/onnx_import_external_data.in.cpp:142]: this test does inference so it requires backend to do so.

[the remark this answers is not in the document]
Comment (t-jankowski) [reply src/frontends/onnx/tests/onnx_import_external_data.in.cpp:142]: I meant that it need not be run on various backends, one (doesn't matter which) should be enough as the purpose of the test is to check proper model external data location propagation. And that is not dependent to a backend.

**Review by t-jankowski — APPROVED** [review]

File: src/frontends/onnx/tests/onnx_import_external_data.in.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (t-jankowski) [reply src/frontends/onnx/tests/onnx_import_external_data.in.cpp:142]: Let's leave it for now - other tests in the file needs to reviewed for the same reason.

**Pull request merged**